ls1intum/Artemis
# `General`: Improve the handling of file extensions

### Checklist
#### General


#### Server
#### Client

### Motivation and Context



In #5328 support for new file extensions was added. However, the associated icons for attachment units were not updated. While working on that issue, I found some more potential for improvements.

### Description


Foremost, this PR adds icons for the new file extensions in attachment units.
It also allows users to upload some more file types: `gif, tar, txt, md, htm, html, json, odc, odi`.
Some other improvements include the removal of code duplication in `file-uploader.service.ts` and displaying the allowed file types also on the _lecture attachments_ form (previously only displayed in the _attachment units_ form).
Finally, the markdown editor has its own list of allowed file types (only images and PDFs), this restriction is now also enforced server-side.

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Lecture

1. Log in to Artemis
2. Navigate to Course Administration
3. Create some _attachment units_ with different file types for the lecture
4. Try to submit the form without selecting a file or selecting disallowed file types
5. Take a look at the created units and ensure a corresponding icon is shown for the file type
6. Upload some _attachments_ for the lecture (should show allowed file types in tooltip)
7. Try to submit the form without selecting a file or selecting disallowed file types
8. Ensure uploading files to the markdown editor still works (e.g., in the exercise problem description)
9. Also try to upload disallowed file types to the markdown editor

### Review Progress





#### Code Review
#### Manual Tests

### Test Coverage


| Class/File | Branches | Lines/Instructions |
|------------|-------:|-----:|
| lecture-attachments.component.ts | 79% | 97% |
| attachment-unit-form.component.ts | 78% | 91% |
| FileResource.java | 78% | 92% |

### Screenshots



<img width="800" alt="Screenshot 2022-08-08 at 13 35 39" src="https://user-images.githubusercontent.com/102237363/183409351-fe02751f-446f-4201-a37a-e33e651f4a3e.png">

<img width="400" alt="Screenshot 2022-08-08 at 13 22 57" src="https://user-images.githubusercontent.com/102237363/183407204-373402c6-461c-47fc-9ef9-c0ba5ecd7c67.png"><img width="400" alt="Screenshot 2022-08-08 at 13 24 07" src="https://user-images.githubusercontent.com/102237363/183407355-85fcf1d1-940c-41f1-9161-949319d03881.png">
